quickwit/quickwit-storage/src/cache/byte_range_cache.rs

```diff
@@ -0,0 +1,359 @@
+            .cache
+            .range(first_matching_block..=last_matching_block)
+            .map(|(k, v)| k.range_start..v.range_end)
+            .collect::<Vec<_>>();
+
+        let can_drop_first = overlapping
+            .first()
+            .map(|r| byte_range.start <= r.start)
+            .unwrap_or(true);
+
+        let can_drop_last = overlapping
+            .last()
+            .map(|r| byte_range.end >= r.end)
+            .unwrap_or(true);
+
+        let (final_range, final_bytes) = if can_drop_first && can_drop_last {
+            (byte_range, bytes)
+        } else {
+            let start = if can_drop_first {
+                byte_range.start
+            } else {
+                // if no first, can_drop_first is false
+                overlapping.first().unwrap().start
+            };
+
+            let end = if can_drop_last {
+                byte_range.end
+            } else {
+                // if no last, can_drop_last is false
+                overlapping.last().unwrap().end
+            };
+
+            let mut buffer = Vec::with_capacity(end - start);
+
+            if !can_drop_first {
+                let first_range = overlapping.first().unwrap();
+                let key = CacheKey::from_borrowed(tag.borrow(), first_range.start);
+                let block = self.cache.get(&key).unwrap();
+
+                let len = first_range.end.min(byte_range.start) - first_range.start;
+                buffer.extend_from_slice(&block.bytes[..len]);
+            }
+
+            buffer.extend_from_slice(&bytes);
+
+            if !can_drop_last {
+                let last_range = overlapping.last().unwrap();
+                let key = CacheKey::from_borrowed(tag.borrow(), last_range.start);
+                let block = self.cache.get(&key).unwrap();
+
+                let start = last_range.start.max(byte_range.end) - last_range.start;
+                buffer.extend_from_slice(&block.bytes[start..]);
+            }
+
+            debug_assert_eq!(end - start, buffer.len());
+
+            (start..end, OwnedBytes::new(buffer))
+        };
+
+        // not sure why, but the borrow check gets unhappy if I create a borrowed
+        // in the loop. It works with .get() instead of .remove() (?).
+        let mut key = CacheKey::from_owned(tag, 0);
+        for range in overlapping.into_iter() {
+            key.range_start = range.start;
+            self.cache.remove(&key);
+            self.drop_item(range.end - range.start);
+        }
+
+        key.range_start = final_range.start;
+        let value = CacheValue {
+            range_end: final_range.end,
+            bytes: final_bytes,
+        };
+        self.cache.insert(key, value);
+        self.record_item(final_range.end - final_range.start);
+    }
+
+    // Return a block that contain everything between query.range_start and range_end
+    fn get_block<'a, 'b: 'a>(
+        &'a self,
+        query: &CacheKey<'b, T>,
+        range_end: usize,
```

Comment:
I could be wrong... because it requires a bit of code wrangling to check that but....

wouldn't the code be clearer by changing this function into...

```
fn get_block<'a>(&'a self, tag: &'b T, byte_range: Range<usize>) -> Option<(Range<usize>, &CacheValue)> {}
```

the goal being to restrict the number of place where we manipulate `CachedKey`.

Reply:
... 
i'm not sure the refactoring I suggest is a good idea, due to the write usage.